tests/PHPStanTests/nsrt/preg-match.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace PregMatchShapes;
+
+use Composer\Pcre\Preg;
+use function PHPStan\Testing\assertType;
+
+function doMatch(string $s): void
+{
+    if (Preg::match('/Price: /i', $s, $matches)) {
+        assertType('array{string}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+    assertType('array{}|array{string}', $matches);
+
+    if (Preg::match('/Price: (£|€)\d+/', $s, $matches)) {
+        assertType('array{string, string}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+    assertType('array{}|array{string, string}', $matches);
+}
+
+function identicalMatch(string $s): void
+{
+    if (Preg::match('/Price: /i', $s, $matches) === 1) {
+        assertType('array{string}', $matches);
+    } else {
+        assertType('array{}', $matches);
+    }
+    assertType('array{}|array{string}', $matches);
+}
+
+function equalMatch(string $s): void
```

Comment:
to get this test working, I think we need a phpstan-core fix  like https://github.com/phpstan/phpstan-src/commit/d842380a1f07014d045dd42d6ccd4204a85e688a

Reply:
https://github.com/phpstan/phpstan-src/pull/3178